bench: BlockAssembler on a mempool with packages

glozow opened the pull request:
Performance of block template building matters as miners likely want to be able to start mining on a block with transactions asap after a block is found. We would want to know if a mempool PR accidentally caused, for example, a 100x slowdown. An `AssembleBlock()` bench exists, but it operates on a mempool with 101 transactions, each with 0 ancestors or descendants and with the same fee. Adding a bench with a more complex mempool is useful because (1) it's more realistic (2) updating packages can potentially cause the algorithm to take a long time.

glozow committed:
[miner] allow bypassing TestBlockValidity

Allows us to test BlockAssembler on transactions without signatures or
mature coinbases (which is what PopulateMempool creates). Also means
that `TestBlockValidity()` is not included in the bench timing.

glozow committed:
[bench] BlockAssembler with mempool packages

The current BlockAssembler bench only tests on a mempool where all
transactions have 0 ancestors or descendants, which does not exercise
any of the package-handling logic in BlockAssembler

glozow committed:
[test util] randomize fee in PopulateMempool

This makes the contents of the mempool more realistic and iterating by
ancestor feerate order more meaningful. If transactions have varying
feerates, it's also more likely that packages will need to be updated
during block template assembly.

jamesob commented on src/test/util/mining.h:
```diff
@@ -24,7 +24,7 @@ std::vector<std::shared_ptr<CBlock>> CreateBlockChain(size_t total_height, const
 
 /** Prepare a block to be mined */
-std::shared_ptr<CBlock> PrepareBlock(const node::NodeContext&, const CScript& coinbase_scriptPubKey);
+std::shared_ptr<CBlock> PrepareBlock(const node::NodeContext&, const CScript& coinbase_scriptPubKey, bool check);
```
https://github.com/bitcoin/bitcoin/pull/26695/commits/3f550b0887db0007633a26aac1dd80ec4723be55

Commit doesn't compile; looks like the `PrepareBlock()` call in `block_assemble.cpp` needs to be updated:
```cpp
bench/block_assemble.cpp:45:9: error: no matching function for call to 'PrepareBlock'
        PrepareBlock(test_setup->m_node, P2WSH_OP_TRUE);
        ^~~~~~~~~~~~
./test/util/mining.h:27:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
std::shared_ptr<CBlock> PrepareBlock(const node::NodeContext&, const CScript& coinbase_scriptPubKey, bool check);
                        ^
1 error generated.
```

jamesob reviewed (commented):
Approach ACK

Change and idea are good, small issue with the first commit.

jamesob reviewed (approved):
ACK 4c20bee65009bd2c95c2a3426030b042f55007ad ([`jamesob/ackr/26695.2.glozow.bench_blockassembler_on`](https://github.com/jamesob/bitcoin/tree/ackr/26695.2.glozow.bench_blockassembler_on))

Nice! Good thing to have.

Could possibly later populate the mempool with a more stressful load, something
like 10,000 transactions? 

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 4c20bee65009bd2c95c2a3426030b042f55007ad ([`jamesob/ackr/26695.2.glozow.bench_blockassembler_on`](https://github.com/jamesob/bitcoin/tree/ackr/26695.2.glozow.bench_blockassembler_on))

Nice! Good thing to have.

Could possibly later populate the mempool with a more stressful load, something
like 10,000 transactions? 
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmOZx4sACgkQepNdrbLE
TwVdwA//bHGKG97lUP5JPiknjulnwVzPvfPyXJi8azZQ2XoK3T1HEA/kJo31wHrY
2mdRgis+kst/emY7wfu/LAAlc06M8cKt5DuPfEgJIPzkidV8JAePMvw8QpDgbd/K
mTu+6mdeGq5Mspy0eS2ItsQEeC4NvZeQSIIf9yCSIhlnCBmeshqBEydN3tduW6Ot
nr6RKGizFO7yJ1ecWX+UA9eYNUYUz4isLehBgoyTh5hz0IJobWIFRmzVCXUQci0y
S3WxFf4RlGnJpWMOOwd/6WZHAdsFOtZ6T4Nux9HEC7nboxb+pfib47CvqCWc0kG3
4mAsJtxasIna/0HTMFD9gApOteG3NByNDKfOkhG1gZYu8tnRA6wm+61WKX6BgJT9
MHHd4WMlov9ye4YaQQazpV8ApXqbjMbFpOLxSeCn3vad/bXo4IWq2x+rpQZTpLJs
+pPNTvqahtk96njeqZcukbnu+JYuEbst4sSlOGMd29fV52M2lyOICK5Jt0PgjfBR
f/lZXhNdd7+1ydj5XOkKE4RAqaQhEf1FYhvR0u+HmFcIVUsEfIYQkvB4z89j28KI
hFGQD3eFPxFPdxBV32GMWHKQ52KBgvzL0Hbbe4a7D/FSbE8YAnRhFXx3jd4AhlAV
8xCDcN/eAthz45ZsTvvoF6NagwRfoXQoP4jzjzFIsNDGD0CKri0=
=cB3y
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-6.0.11-arch1-1-x86_64-with-glibc2.36

Configured with ./configure --without-bdb 'CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare ' --disable-gui-tests --disable-fuzz-binary --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/bin/clang++ CC=/usr/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare  -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4.1 -msse4.2 -msse4 -msha  i

Compiler version: clang version 14.0.6
```

</p></details>

MarcoFalke commented on src/test/util/mining.cpp:
```diff
@@ -74,10 +74,12 @@ CTxIn MineBlock(const NodeContext& node, const CScript& coinbase_scriptPubKey)
-std::shared_ptr<CBlock> PrepareBlock(const NodeContext& node, const CScript& coinbase_scriptPubKey)
+std::shared_ptr<CBlock> PrepareBlock(const NodeContext& node, const CScript& coinbase_scriptPubKey, bool check)
 {
+    BlockAssembler::Options options;
```
not sure if it matters, but previously this was (implicitly) calling `DefaultOptions`, now it calls `Options`.

glozow replied:
Added an `ApplyArgsManOptions` so we can both manually manipulate the options and get the config from `ArgsManager`.

MarcoFalke commented on src/test/util/mining.cpp:
```diff
@@ -74,10 +74,12 @@ CTxIn MineBlock(const NodeContext& node, const CScript& coinbase_scriptPubKey)
 }
 
-std::shared_ptr<CBlock> PrepareBlock(const NodeContext& node, const CScript& coinbase_scriptPubKey)
+std::shared_ptr<CBlock> PrepareBlock(const NodeContext& node, const CScript& coinbase_scriptPubKey, bool check)
```
nit: Could pass in the options struct to avoid having to change this again if another option is added?
Could also add a redirect function without the argument, similar to BlockAssembler, to avoid having to change existing call sites?

Feel free to ignore, but then you could use clang-tidy named-args to at least document the bool literals.

glozow replied:
Good idea, did this instead of the bool

MarcoFalke commented on src/bench/block_assemble.cpp:
```diff
@@ -45,5 +45,18 @@ static void AssembleBlock(benchmark::Bench& bench)
+    auto testing_setup = MakeNoLogFileContext<TestChain100Setup>();
+    CTxMemPool& pool = *testing_setup.get()->m_node.mempool;
+    LOCK2(cs_main, pool.cs);
+    const auto transactions = testing_setup->PopulateMempool(det_rand, 1000, true);
```
nit: could use clang-tidy named args?

glozow replied:
Added

MarcoFalke commented on src/bench/block_assemble.cpp:
```diff
@@ -45,5 +45,18 @@ static void AssembleBlock(benchmark::Bench& bench)
+    FastRandomContext det_rand{true};
+    auto testing_setup = MakeNoLogFileContext<TestChain100Setup>();
+    CTxMemPool& pool = *testing_setup.get()->m_node.mempool;
+    LOCK2(cs_main, pool.cs);
```
nit: why?

glozow replied:
Removed, thanks

maflcko reviewed (approved):
lgtm. Left some nits

stickies-v reviewed (commented):
Concept ACK - nice to have this kind of benchmark, will be helpful in e.g. #26289

glozow committed:
[refactor] add helper to apply ArgsManager to BlockAssembler::Options

This allows us to both manually manipulate options and grab values from
ArgsManager (i.e. -blockmaxweight and -blockmintxfee config options)
when constructing BlockAssembler::Options. Prior to this change, the
only way to apply the config options is by ctoring BlockAssembler with
no options, which calls DefaultOptions().

glozow committed:
[refactor] parameterize BlockAssembler::Options in PrepareBlock

glozow commented on src/bench/block_assemble.cpp:
```diff
@@ -42,8 +42,21 @@ static void AssembleBlock(benchmark::Bench& bench)
+    auto testing_setup = MakeNoLogFileContext<TestChain100Setup>();
+    CTxMemPool& pool = *testing_setup.get()->m_node.mempool;
+    LOCK2(cs_main, pool.cs);
+    const auto transactions = testing_setup->PopulateMempool(det_rand, 1000, true);
```
Some taken, most lines deleted

glozow commented on src/test/util/mining.cpp:
```diff
@@ -61,7 +61,7 @@ std::vector<std::shared_ptr<CBlock>> CreateBlockChain(size_t total_height, const
 CTxIn MineBlock(const NodeContext& node, const CScript& coinbase_scriptPubKey)
 {
-    auto block = PrepareBlock(node, coinbase_scriptPubKey);
+    auto block = PrepareBlock(node, coinbase_scriptPubKey, true);
```
Line is removed now

glozow commented on src/test/util/setup_common.cpp:
```diff
@@ -397,15 +397,15 @@ std::vector<CTransactionRef> TestChain100Setup::PopulateMempool(FastRandomContex
         const size_t num_outputs = det_rand.randrange(24) + 1;
-        // Approximately 1000sat "fee," equal output amounts.
-        const CAmount amount_per_output = (total_in - 1000) / num_outputs;
+        const CAmount fee = 100 * det_rand.randrange(30);
```
Updated commit message.
This is a narrowing conversion - randrange returns uint64_t, CAmount is a 32.

glozow commented on src/node/miner.cpp:
```diff
@@ -170,7 +173,7 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
 
     BlockValidationState state;
-    if (!TestBlockValidity(state, chainparams, m_chainstate, *pblock, pindexPrev, GetAdjustedTime, false, false)) {
+    if (test_block_validity && !TestBlockValidity(state, chainparams, m_chainstate, *pblock, pindexPrev, GetAdjustedTime, false, false)) {
```
Done

glozow committed:
[test util] lock cs_main before pool.cs in PopulateMempool

jamesob reviewed (approved):
ACK 628ce92c494c6d3b6181d2ee6f64494de38a67e3 ([`jamesob/ackr/26695.3.glozow.bench_blockassembler_on`](https://github.com/jamesob/bitcoin/tree/ackr/26695.3.glozow.bench_blockassembler_on))

re-ACK

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 628ce92c494c6d3b6181d2ee6f64494de38a67e3 ([`jamesob/ackr/26695.3.glozow.bench_blockassembler_on`](https://github.com/jamesob/bitcoin/tree/ackr/26695.3.glozow.bench_blockassembler_on))

re-ACK
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmOcpWIACgkQepNdrbLE
TwW/5w/+MpLKrQ5eLWc7uXQ2P1ML82Q1EKDXzGC4bmBFZZ2t7Ynj78enPUf1ukD2
o8hT28touxw5XSWF2Zcwjhh+DRWU6sqYJInpOXhW1366iqS2DcvDQZOctwEEUasO
RM6lcRfNNvz4FyuiV4Cfn/liVHK6ipCQoPY88mjyy/I32XzwB6wXKNopRCV8GfFV
pIU10BM9tVobwHJhFRvTQve3nhwXjIbwJ0Vn9peUqKRSmuxDkFf2R6l/3YUwEFjk
iPJQ+SW4fo0h6bDThZ8GOfT8F4u52cnsx97zOck9f6wmehQoKUDIj6s2Epa/n/0q
WZXx6+xI2RS3GuS7fM9et1+0H8TmyHhkvSpzkSiOg+npWBqPl+8tCgqQftUS4s+D
HY/BoP1arD2nHsQ/yBksymASmnFKMyR0RWz8r8JheUY3gGXRDCQZkUCkVMMjpV3O
tOIx6k1h8dfyaJqxW2e6o3r2dNho79Ap2nQS7UnRfFu80KebgyV0SZYusK3+OT4a
Wkm4VBO3l36BQJwOBmu6nQ3mpbgF4TvQOhj9F65GUaLrsAKrNn/YrcnV2U+g/RmS
JaXsSUio7DukkYccLLLOh9WvYOXkxohIXNA7DBpk+VES8PBDN5Q5ZBuCmJSRt0HC
VwSxSbNA4Pedk6pmaebOAgTUcQxY9yJOaBQLEqyhR3SRE4YgrVQ=
=rMYG
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-6.0.11-arch1-1-x86_64-with-glibc2.36

Configured with ./configure --without-bdb 'CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare ' --disable-gui-tests --disable-fuzz-binary --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/bin/clang++ CC=/usr/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare  -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4.1 -msse4.2 -msse4 -msha  i

Compiler version: clang version 14.0.6
```

</p></details>

jamesob reviewed (approved):
ACK b05347204d2d1cad030fc794aae2b0d29c35db7d ([`jamesob/ackr/26695.4.glozow.bench_blockassembler_on`](https://github.com/jamesob/bitcoin/tree/ackr/26695.4.glozow.bench_blockassembler_on))

Oops, previous ACK already stale. New ACK incorporates

```diff
-            m_node.mempool->addUnchecked(CTxMemPoolEntry(ptx, 1000, 0, 1, false, 4, lp));
+            m_node.mempool->addUnchecked(CTxMemPoolEntry(ptx, /*fee=*/(total_in - num_outputs * amount_per_output),
+                                                         /*time=*/0, /*entry_height=*/1,
+                                                         /*spends_coinbase=*/false, /*sigops_cost=*/4, lp));
         }
         --num_transactions;
     }
```

glozow commented:
Yes apologies @jamesob, I noticed I was making the tx fee random but still setting the mempool entry fee to 1000. Thanks for such a quick review!

MarcoFalke commented on src/bench/block_assemble.cpp:
```diff
@@ -45,5 +46,18 @@ static void AssembleBlock(benchmark::Bench& bench)
+    auto testing_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    testing_setup->PopulateMempool(det_rand, /*num_transactions=*/1000, /*submit=*/true);
+    node::BlockAssembler::Options assembler_options;
+    assembler_options.test_block_validity = false;
```
Is there any value in disabling it? Sure, the benchmark will be less "micro", but at the same time it would be missing if a "package" block is slowing down CNB via TBV. Or is taking TBV a lot longer than the remainder of CNB, which would make this benchmark ineffective?

glozow replied:
One reason is TBV taking time, another is TBV would fail because `PopulateMempool()` doesn't sign the txns and spends immature coinbases. I can change it to generate keys and mine 100 blocks in between creating and submitting the transactions if people want.

kevkevinpal commented on src/bench/block_assemble.cpp:
```diff
@@ -42,8 +42,21 @@ static void AssembleBlock(benchmark::Bench& bench)
 }
 
 BENCHMARK(AssembleBlock, benchmark::PriorityLevel::HIGH);
+BENCHMARK(BlockAssemblerAddPackageTxns, benchmark::PriorityLevel::HIGH);
```
this is what I get when trying to run the `BlockAssemblerAddPackageTxns` benchmark

Specs:
```
mac book pro 2019
2.3 GHz 8-Core Intel Core i9
16 GB 2667 MHz DDR4
AMD Radeon Pro 5500M 4 GB
```

Results:
```
./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns --sanity-check
Running with --sanity-check option, benchmark results will be useless.

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      813,714,108.00 |                1.23 |    0.0% |      0.81 | `BlockAssemblerAddPackageTxns`
```

Without `--sanity-check`:
```
./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      830,079,779.00 |                1.20 |    1.4% |      9.16 | `BlockAssemblerAddPackageTxns`
```

achow101 replied:
When compiled with `--enable-debug`, this slows down `make check` by ~30s for me. When run without `--sanity-check`, it takes ~280s. I think this needs to be a low priority benchmark.

stickies-v replied:
Yeah sorry, we had medium in the initial PR but it was removed later on, so agreed with changing it to low. We haven't defined any guidelines on when to downgrade to low afaik, but since devs run `make check` often, I think adding benches with meaningful (e.g. >1s?) `--sanity-check` overhead should run in low priority - unless their complexity/importance warrants otherwise.

kevkevinpal commented:
I saw in 50218324dac18556d87688dc1a8e89bbe4d5f69e you change 
`LOCK2(m_node.mempool->cs, cs_main);` -> `LOCK2(cs_main, m_node.mempool->cs);`
I'm just curious as to the reason

glozow commented:
> I saw in [5021832](https://github.com/bitcoin/bitcoin/commit/50218324dac18556d87688dc1a8e89bbe4d5f69e) you change `LOCK2(m_node.mempool->cs, cs_main);` -> `LOCK2(cs_main, m_node.mempool->cs);` I'm just curious as to the reason

It's an inconsistent order of acquiring mutexes, which along with other things can cause deadlock. I missed it when first implementing `PopulateMempool()` - until this PR, it was only called with both locks already acquired. Compiling with `--enable-debug` / `-DDEBUG_LOCKORDER` found it.

Emzy commented:
Tested ACK b05347204d2d1cad030fc794aae2b0d29c35db7d

this is what I get when running the ```BlockAssemblerAddPackageTxns``` benchmark

Specs:
mac book pro 2019
2,6 GHz 6-Core Intel Core i7
16 GB 2667 MHz DDR4
macos 13.0.1 (22A400)

```
$ ./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns --sanity-check
Running with --sanity-check option, benchmark results will be useless.

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      777,853,228.00 |                1.29 |    0.0% |      0.78 | `BlockAssemblerAddPackageTxns`
```
Without --sanity-check:
```
$ ./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns 

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      787,270,151.00 |                1.27 |    0.8% |      8.94 | `BlockAssemblerAddPackageTxns`
```

Specs:
Thin Client 
2,6 GHz 2-Core AMD G-T56N Processor
4 GB 
Ubuntu 20.04.5 LTS

```
$ ./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns --sanity-check
Running with --sanity-check option, benchmark results will be useless.
Warning, results might be unstable:
* CPU frequency scaling enabled: CPU 0 between 825.0 and 1,650.0 MHz
* CPU governor is 'ondemand' but should be 'performance'
* Turbo is enabled, CPU frequency will fluctuate

Recommendations
* Use 'pyperf system tune' before benchmarking. See https://github.com/psf/pyperf

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    3,488,170,147.00 |                0.29 |    0.0% |      3.49 | `BlockAssemblerAddPackageTxns`
```
Without --sanity-check:
```
$ ./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns
Warning, results might be unstable:
* CPU frequency scaling enabled: CPU 0 between 825.0 and 1,650.0 MHz
* CPU governor is 'ondemand' but should be 'performance'
* Turbo is enabled, CPU frequency will fluctuate

Recommendations
* Use 'pyperf system tune' before benchmarking. See https://github.com/psf/pyperf

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    3,477,761,468.00 |                0.29 |    0.0% |     38.30 | `BlockAssemblerAddPackageTxns`

```

hernanmarino reviewed (commented):
ACK b05347204d2d1cad030fc794aae2b0d29c35db7d . Tested / bench'ed correctly

kevkevinpal commented:
Tested ACK [b053472](https://github.com/bitcoin/bitcoin/pull/26695/commits/b05347204d2d1cad030fc794aae2b0d29c35db7d)

left some test results here
https://github.com/bitcoin/bitcoin/pull/26695#discussion_r1053907539

pablomartin4btc commented on src/node/miner.cpp:
```diff
@@ -72,18 +74,22 @@ BlockAssembler::BlockAssembler(Chainstate& chainstate, const CTxMemPool* mempool
         options.blockMinFeeRate = CFeeRate{DEFAULT_BLOCK_MIN_TX_FEE};
     }
+}
+static BlockAssembler::Options DefaultOptions()
```
```suggestion
static BlockAssembler::Options CustomOptions()
```
nit: as discussed on the [PR review club](https://bitcoincore.reviews/26695), perhaps the `"DefaultOptions"` could be renamed to avoid confusion (`"CustomOptions"` could be an alternative), unless there's a reason for not doing it.

pablomartin4btc reviewed (commented):
Tested ACK. 
I've executed the bench successfully; compiled with `--enable-debug`.

stickies-v reviewed (approved):
ACK 04528054f

I think [this comment](https://github.com/bitcoin/bitcoin/pull/26695/files#r1048480684) can be resolved in a follow-up if preferable but I'd prefer the [`gArgs` parameter be renamed](https://github.com/bitcoin/bitcoin/pull/26695#discussion_r1054562437) in this PR already since it's quite confusing.

kevkevinpal commented:
Tested ACK [0452805](https://github.com/bitcoin/bitcoin/pull/26695/commits/04528054fcde61aa00e009dbbe1ac350ca1cf748)
Compiled this time using `--enable-debug`

Specs:

mac book pro 2019
2.3 GHz 8-Core Intel Core i9
16 GB 2667 MHz DDR4
AMD Radeon Pro 5500M 4 GB


`➜  bitcoin git:(PR26695) ✗ ./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns --sanity-check`
```
Running with --sanity-check option, benchmark results will be useless.
Warning, results might be unstable:
* DEBUG defined

Recommendations
* Make sure you compile for Release

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|   45,742,309,331.00 |                0.02 |    0.0% |     45.74 | `BlockAssemblerAddPackageTxns`
```
`➜  bitcoin git:(PR26695) ✗ ./src/bench/bench_bitcoin -filter=BlockAssemblerAddPackageTxns`

```
Warning, results might be unstable:
* DEBUG defined

Recommendations
* Make sure you compile for Release

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|   43,927,650,019.00 |                0.02 |    0.7% |    488.03 | `BlockAssemblerAddPackageTxns`
```

achow101 commented:
ACK 04528054fcde61aa00e009dbbe1ac350ca1cf748

stickies-v commented:
> I think [this comment](https://github.com/bitcoin/bitcoin/pull/26695/files#r1048480684) can be resolved in a follow-up if preferable but I'd prefer the [`gArgs` parameter be renamed](https://github.com/bitcoin/bitcoin/pull/26695#discussion_r1054562437) in this PR already since it's quite confusing.

Follow-ups in https://github.com/bitcoin/bitcoin/pull/26883
